client/state/guided-tours/selectors/index.js

```diff
@@ -22,35 +21,37 @@ const SECTIONS_WITHOUT_TOURS = [
 
 const debug = debugFactory( 'calypso:guided-tours' );
 
-const mappable = ( x ) => ( ! Array.isArray( x ) ? [ x ] : x );
+function tourMatchesPath( tour, path ) {
+	if ( ! tour.path ) {
+		return false;
+	}
 
-const relevantFeatures = flatMap( GuidedToursConfig, ( tourMeta, key ) =>
-	mappable( tourMeta.path ).map( ( path ) => ( {
-		tour: key,
-		when: tourMeta.when,
-		path,
-	} ) )
-);
+	if ( Array.isArray( tour.path ) ) {
+		return tour.path.some( ( p ) => path.startsWith( p ) );
+	}
+
+	return path.startsWith( tour.path );
+}
 
 /*
  * Returns a collection of tour names. These tours are selected if the user has
  * recently navigated to a section of Calypso that comes with a corresponding
  * tour.
  */
 const getToursFromFeaturesReached = createSelector(
-	( state ) => [
-		...new Set(
-			getActionLog( state )
-				.filter( ( { type } ) => type === ROUTE_SET )
-				.reduceRight( ( allTours, { path: triggerPath } ) => {
-					const newTours = relevantFeatures
-						.filter( ( { path: featurePath } ) => startsWith( triggerPath, featurePath ) )
-						.map( ( feature ) => feature.tour );
-
-					return newTours ? [ ...allTours, ...newTours ] : allTours;
-				}, [] )
-		),
-	],
+	( state ) => {
+		// list of recent navigations in reverse order
+		const navigationActions = getActionLog( state )
+			.filter( ( { type } ) => type === ROUTE_SET )
+			.reverse();
+		// find tours that match by route path
+		const tourEntries = Object.entries( GuidedToursConfig );
```

Comment:
Should this be moved out of the selector if it's the same on every call?

Reply:
In d181248940 I implemented an idea that I think is even better: convert `GuidedToursConfig` to an array. Instead of a map of configs indexed by `name`, it's now an array which you need to search with `config.find( tour => tour.name === name )`. It simplifies the code, and as the config is an array with 11 items, we don't need to worry about performance at all.